source/common/config/delta_subscription_state.cc

```diff
@@ -202,6 +315,39 @@ DeltaSubscriptionState::getNextRequestAckless() {
   return request;
 }
 
+bool DeltaSubscriptionState::isInitialRequestForLegacyWildcard() {
+  if (in_initial_legacy_wildcard_) {
+    requested_resource_state_.insert_or_assign(Wildcard, ResourceState::waitingForServer());
```

Comment:
is this PR preparation for on-demand CDS? 

does it mean that on-demand for wildcard XDSes will always call NewGrpcMuxImpl::updateWatch with a "*" in it? otherwise the updateWatch "degrades" the stream into a "non-wildcard" stream.

Reply:
> is this PR preparation for on-demand CDS?

Yes, on-demand CDS requires these changes. But the same would be for mixing on-demand requests with wildcard requests for any resource, really.

> 
> does it mean that on-demand for wildcard XDSes will always call NewGrpcMuxImpl::updateWatch with a "*" in it? otherwise the updateWatch "degrades" the stream into a "non-wildcard" stream.

Reply:
> does it mean that on-demand for wildcard XDSes will always call NewGrpcMuxImpl::updateWatch with a "*" in it? otherwise the updateWatch "degrades" the stream into a "non-wildcard" stream.

Not sure, I'd say it depends on the implementation of xDS. I'd say that xDSes initially end up calling `NewGrpcMuxImpl::updateWatch`, but then they should be calling `NewGrpcMuxImpl::requestOnDemandUpdate` when a new resource is requested on demand.

Maybe the useful way of thinking here is not about stream being "wildcard" or "non-wildcard", but rather about `*` being a resource. And sending an empty initial request on the stream as a special case of sending just `*`.

Reply:
so the purpose is to "turn a non-wildcard XDS into a wildcard XDS by explicitly requesting *"? 

aside from that, I think with this PR, is this a path to "turn wildcard into non-wildcard"? 

1. Client starts CDS, it's wildcard
2. Server returns A, B, C
3. Client adds explicit request for "D", but the connection is now disconnected
4. reconnection happens, but the in_initial_legacy_wildcard_ is false now that it sees a non-"Wildcard" resource, names_added_ will be non-empty, isInitialRequestForLegacyWildcard will be false now, the "*" is not inserted by isInitialRequestForLegacyWildcard. 
5. request sent to Server, with non-empty "resource_names_subscribe", Server sees a first request with non-empty list of resources, it will treat it as a "non-wildcard" and only push A,B,C,D updates now. 

Is this desired?

Reply:
Not commenting on the implementation here, only on the expected behavior in this scenario:

In step 1, the client is subscribed only to the wildcard.  In step 3, the client adds an additional subscription to D, so it is now subscribed to both the wildcard and to D.  The disconnection does not affect the client's set of subscriptions, so when it reconnects, it should send a message with `resource_names_subscribe` set to both `*` and D.

The fact that the server sends resources A, B, and C in response to the wildcard subscription does not change the client's subscriptions; the client is still subscribed to the wildcard, not to A, B, and C individually.

Reply:
> so the purpose is to "turn a non-wildcard XDS into a wildcard XDS by explicitly requesting *"?
> 
> aside from that, I think with this PR, is this a path to "turn wildcard into non-wildcard"?
> 
>     1. Client starts CDS, it's wildcard
> 
>     2. Server returns A, B, C

If this happens, it means that we have received a reply to our the initial request on the stream. When creating the initial request, we have inserted `*` to `requested_resource_state_` in `isInitialRequestForLegacyWildcard`.

> 
>     3. Client adds explicit request for "D", but the connection is now disconnected
> 
>     4. reconnection happens, but the in_initial_legacy_wildcard_ is false now that it sees a non-"Wildcard" resource, names_added_ will be non-empty, isInitialRequestForLegacyWildcard will be false now, the "*" is not inserted by isInitialRequestForLegacyWildcard.

This is true, but `*` is already inserted into `requested_resource_state_` as described above.

> 
>     5. request sent to Server, with non-empty "resource_names_subscribe", Server sees a first request with non-empty list of resources, it will treat it as a "non-wildcard" and only push A,B,C,D updates now.

The server will see `*` and `D` as all the resources from `requested_resource_state_` will be added to the request.

> 
> 
> Is this desired?